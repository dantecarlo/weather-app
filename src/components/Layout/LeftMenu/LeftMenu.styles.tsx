import { Box, Link } from '@mui/material'
import { styled } from '@mui/material/styles'
import { OpenInNew as OpenInNewIcon } from '@mui/icons-material'

export const Root = styled(Box)({
  marginLeft: 'auto'
})

export const StyledOpenInNewIcon = styled(OpenInNewIcon)({
  fontSize: '1rem !important',
  bottom: -3,
  position: 'relative'
})

export const RightBox = styled(Box)(({ theme }) => ({
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  justifyContent: 'flex-end',
  [theme.breakpoints.down('sm')]: {
    display: 'none'
  }
}))

export const RightBoxLink = styled(Link)(({ theme }) => ({
  color: `${theme.palette.common.white}9 !important`,
  marginLeft: theme.spacing(1),
  marginRight: theme.spacing(1),
  paddingLeft: theme.spacing(1),
  paddingRight: theme.spacing(1),
  '&:hover': {
    color: `${theme.palette.common.white} !important`
  },
  '&:last-child': {
    marginRight: `${theme.spacing(0)} !important`,
    paddingRight: theme.spacing(0)
  }
}))
