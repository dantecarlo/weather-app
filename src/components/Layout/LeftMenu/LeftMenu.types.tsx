export interface IExternalLinks {
  displayName: string
  href: string
  external: boolean
}

export interface ILeftMenu {
  linkList: IExternalLinks[]
}
