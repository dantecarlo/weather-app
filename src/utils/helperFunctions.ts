// eslint-disable-next-line import/prefer-default-export
export function groupBy<T extends Record<string, any>, K extends keyof T>(
  array: T[],
  key: K | { (obj: T): string }
): Record<string, T[]> {
  const keyFn = key instanceof Function ? key : (obj: T) => obj[key]
  return array.reduce((objectsByKeyValue, obj) => {
    const value = keyFn(obj)
    // eslint-disable-next-line no-param-reassign
    objectsByKeyValue[value] = [...(objectsByKeyValue[value] || []), obj]
    return objectsByKeyValue
  }, {} as Record<string, T[]>)
}
