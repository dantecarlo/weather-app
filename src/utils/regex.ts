/*
 * name validation
 * accepted: letters & spaces, minimum 3 chars, maximum 15 chars
 */
export const inputTextRegex = /^[a-zA-ZA-zÀ-ÿ'ç\- ]+$/gm
export const alphanumericRegex = /^[a-zA-Z0-9 ]*$/gm

/*
 * email validation
 */
// eslint-disable-next-line operator-linebreak
export const emailRegex =
  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/g

/*
 * password validation, should contain:
 * (?=.*\d): at least one digit
 * (?=.*[a-z]): at least one lower case
 * (?=.*[A-Z]): at least one uppercase case
 * [0-9a-zA-Z]{6,}: at least 6 from the mentioned characters
 */
export const passwordRegex = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[0-9a-zA-Z]{6,}$/
