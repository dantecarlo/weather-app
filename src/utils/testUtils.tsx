import { FC, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import i18next from 'i18next'
import { I18nextProvider } from 'react-i18next'
import { BrowserRouter } from 'react-router-dom'

import en from '../i18n/en.json'
import es from '../i18n/es.json'
import theme from '../theme'
import { ContextProvider } from '../contexts/ContextProvider'

i18next.init({
  fallbackLng: 'enEN',
  interpolation: {
    escapeValue: false
  },
  lng: 'enEN',
  resources: {
    enEN: { global: en },
    esES: { global: es }
  }
})

type Props = {
  children?: React.ReactNode
}

const AllTheProviders: FC<Props> = ({ children }) => (
  <BrowserRouter>
    <I18nextProvider i18n={i18next}>
      <ThemeProvider theme={theme}>
        <ContextProvider>{children}</ContextProvider>
      </ThemeProvider>
    </I18nextProvider>
  </BrowserRouter>
)

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options
  })

AllTheProviders.defaultProps = {
  children: []
}

export * from '@testing-library/react'

export { customRender as render }
