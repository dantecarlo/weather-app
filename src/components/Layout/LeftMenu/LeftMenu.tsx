import { FC } from 'react'

import { RightBox, RightBoxLink, Root, StyledOpenInNewIcon } from './LeftMenu.styles'

interface IExternalLinks {
  displayName: string
  href: string
  external: boolean
}

interface ILeftMenu {
  linkList: IExternalLinks[]
}

const LeftMenu: FC<ILeftMenu> = (props) => {
  const { linkList } = props

  return (
    <Root>
      <RightBox>
        {linkList.map(({ displayName, href, external }, index) => {
          const linkKey = `link-key-${displayName}-${index}`

          return (
            <RightBoxLink
              key={linkKey}
              href={href}
              rel="noopener"
              target="_blank"
              underline="none"
              variant="button"
            >
              {displayName} {external && <StyledOpenInNewIcon />}
            </RightBoxLink>
          )
        })}
      </RightBox>
    </Root>
  )
}

export default LeftMenu
