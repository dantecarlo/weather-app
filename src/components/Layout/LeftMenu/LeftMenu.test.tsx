import { RenderResult } from '@testing-library/react'
import { get } from 'lodash'
import { render } from 'utils/testUtils'
import LeftMenu from './LeftMenu'

describe('<LeftMenu>', () => {
  let component: RenderResult

  const linkTestList = [
    {
      displayName: 'test1',
      href: '#test1',
      external: true
    },
    {
      displayName: 'test2',
      href: '#test2',
      external: true
    },
    {
      displayName: 'test3',
      href: '#test3',
      external: true
    }
  ]

  beforeEach(() => {
    component = render(<LeftMenu linkList={linkTestList} />)
  })

  test('renders correctly', () => {
    const { asFragment } = component

    expect(asFragment()).toMatchSnapshot()
  })

  test('render all links', () => {
    const { container } = component

    linkTestList.forEach((link) => {
      expect(container.innerHTML).toMatch(get(link, 'displayName'))
    })
  })
})
