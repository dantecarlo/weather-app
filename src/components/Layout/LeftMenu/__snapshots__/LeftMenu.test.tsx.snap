// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LeftMenu> renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin-left: auto;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

@media (max-width:599.95px) {
  .emotion-1 {
    display: none;
  }
}

.emotion-2 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  color: #26225e;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff9!important;
  margin-left: 8px;
  margin-right: 8px;
  padding-left: 8px;
  padding-right: 8px;
}

.emotion-2:hover {
  color: #fff!important;
}

.emotion-2:last-child {
  margin-right: 0px!important;
  padding-right: 0px;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  font-size: 1rem!important;
  bottom: -3px;
  position: relative;
}

<div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <a
        class="MuiTypography-root MuiTypography-button MuiLink-root MuiLink-underlineNone emotion-2"
        href="#test1"
        rel="noopener"
        target="_blank"
      >
        test1 
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="OpenInNewIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          />
        </svg>
      </a>
      <a
        class="MuiTypography-root MuiTypography-button MuiLink-root MuiLink-underlineNone emotion-2"
        href="#test2"
        rel="noopener"
        target="_blank"
      >
        test2 
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="OpenInNewIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          />
        </svg>
      </a>
      <a
        class="MuiTypography-root MuiTypography-button MuiLink-root MuiLink-underlineNone emotion-2"
        href="#test3"
        rel="noopener"
        target="_blank"
      >
        test3 
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="OpenInNewIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
